File: python/src/sort/recursion_insertion_sort.py

def recursion_insertion_sort(in_array):
    if len(in_array) < 2:
        return in_array

    last   = in_array[-1]
    rest   = recursion_insertion_sort( in_array[:-1] )
    return merge_recursion_insertion_sort( rest, last )

def merge_recursion_insertion_sort( rest, last ):
    for i in range( len(rest) ):
        if rest[i] > last:
            return rest[0:i] + [last] + rest[i:]
    return rest + [last]

File: c#/Sort/RecursionInsertionSort.cs
using System.Collections;

namespace Sort
{
    public class RecursionInsertionSort : ISort
    {
        public ArrayList Sort(ArrayList inArray)
        {
            if (inArray.Count < 2)
                return inArray;
            
            int index = inArray.Count - 1;
            var last = (int)inArray[index];
            inArray.RemoveAt(index);
            ArrayList rest = Sort(inArray);

            return Merge(rest, last);
        }

        public ArrayList Merge(ArrayList rest, int last)
        {
            for (int i = 0; i < rest.Count; i++ )
            {
                if ((int)rest[i] > last)
                {
                    rest.Insert(i, last);
                    return rest;
                }
            }

            rest.Add(last);
            return rest;
        }
    }
}

